src/Framework/Assert.php

```diff
@@ -819,6 +819,8 @@ public static function assertAttributeNotEquals($expected, string $actualAttribu
      *
      * @throws ExpectationFailedException
      * @throws \SebastianBergmann\RecursionContext\InvalidArgumentException
+     *
+     * @psalm-assert TEmptyMixed $actual
```

Comment:
All of this file will need at least 2 tests for each touched assertion, to be picked from the following:

 * [ ] positive test (no issues)
 * [ ] negative test (impossible type)
 * [ ] redundant test (assertion always passes)

Reply:
I think this should be `@psalm-assert empty $actual`

Reply:
I'll flesh these out one by one with tests as I get to work on them. I used the various `T*` from the plugin documentation, so that may be incorrect.

Reply:
Oh that's _definitely_ incorrect. Where's that?

Reply:
https://github.com/vimeo/psalm/blob/3.3.0/docs/plugins_type_system.md

I guess that's post-parsing types

Reply:
Yeah, it's complicated.

`@psalm-assert` takes anything that the `Reconciler` class can understand.

It's not documented anywhere, but it takes any unadorned type (e.g. `int`, `Foo`, `scalar`), and the special case `empty`.

It also takes negations of the same e.g. `!int`, `!Foo`, `!numeric`, `!empty` etc.

Reply:
After inspecting this further, I only attacked the problem from a "correct type" point of view.